src/main/java/org/opensearch/ad/model/Entity.java

```diff
@@ -29,66 +29,146 @@
+     * @param attrs attribute map
+     * @return the created entity
+     */
+    public static Entity createEntityFromOrderedMap(String detectorId, SortedMap<String, String> attrs) {
+        return new Entity(attrs);
+    }
+
+    private Entity(SortedMap<String, String> orderedAttrs) {
+        this.attributes = orderedAttrs;
     }
 
     public Entity(StreamInput input) throws IOException {
-        this.name = input.readString();
-        this.value = input.readString();
+        this.attributes = new TreeMap<>(input.readMap(StreamInput::readString, StreamInput::readString));
     }
 
     @Override
     public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
-        XContentBuilder xContentBuilder = builder.startObject().field(ENTITY_NAME_FIELD, name).field(ENTITY_VALUE_FIELD, value);
-        return xContentBuilder.endObject();
+        builder.startArray();
+        for (Map.Entry<String, String> attr : attributes.entrySet()) {
+            builder.startObject().field(ATTRIBUTE_NAME_FIELD, attr.getKey()).field(ATTRIBUTE_VALUE_FIELD, attr.getValue()).endObject();
+        }
+        builder.endArray();
+        return builder;
     }
 
     public static Entity parse(XContentParser parser) throws IOException {
+        SortedMap<String, String> entities = new TreeMap<>();
         String parsedValue = null;
         String parsedName = null;
 
-        ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser);
-        while (parser.nextToken() != XContentParser.Token.END_OBJECT) {
-            String fieldName = parser.currentName();
-            parser.nextToken();
-
-            switch (fieldName) {
-                case ENTITY_NAME_FIELD:
-                    parsedName = parser.text();
-                    break;
-                case ENTITY_VALUE_FIELD:
-                    parsedValue = parser.text();
-                    break;
-                default:
-                    break;
+        ensureExpectedToken(XContentParser.Token.START_ARRAY, parser.currentToken(), parser);
+        while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
+            ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser);
+            while (parser.nextToken() != Token.END_OBJECT) {
+                String fieldName = parser.currentName();
+                // move to the field value
+                parser.nextToken();
+                switch (fieldName) {
+                    case ATTRIBUTE_NAME_FIELD:
+                        parsedName = parser.text();
+                        break;
+                    case ATTRIBUTE_VALUE_FIELD:
+                        parsedValue = parser.text();
+                        if (parsedName != null && parsedValue != null) {
+                            entities.put(parsedName, parsedValue);
+                        }
+                        parsedValue = null;
+                        parsedName = null;
```

Comment:
Nit - why have this logic embedded in the `ATTRIBUTE_VALUE_FIELD` case? Can it be moved after the switch statement for readability? I see this can be more efficient if it is always guaranteed that the attribute name will come before the value, just wonder if it adds confusion to add this here.

Reply:
If we moved this after the switch, it would reset for every field.  Here, I want to reset every time I have seen a name-value pair.

Reply:
I was thinking after the switch, but still within the `while` loop, so it would only reset after `parsedName` and `parsedField` were populated. Seems ok to leave as-is, just wanting to clarify

Reply:
Seems this logic can't work if `ATTRIBUTE_VALUE_FIELD` is before `ATTRIBUTE_NAME_FIELD`.  @ohltyler , do you mean change to this
```
        // Iterate every entity attribute pair
        while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
            ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser);
            
            String parsedValue = null;
            String parsedName = null;
            // parse current entity attribute pair
            while (parser.nextToken() != Token.END_OBJECT) {
                String fieldName = parser.currentName();
                // move to the field value
                parser.nextToken();
                switch (fieldName) {
                    case ATTRIBUTE_NAME_FIELD:
                        parsedName = parser.text();
                        break;
                    case ATTRIBUTE_VALUE_FIELD:
                        parsedValue = parser.text();
                    default:
                        break;
            }

            if (parsedName != null && parsedValue != null) {
                entities.put(parsedName, parsedValue);
            }
        }
```

Reply:
@ylwu-amzn almost, except still clear the `parsedValue` and `parsedName` in the `if` block

Reply:
right.  Changed.

Reply:
@ylwu-amzn I was thinking within the inner `while` loop still, like this:
```
        // Iterate every entity attribute pair
        while (parser.nextToken() != XContentParser.Token.END_ARRAY) {
            ensureExpectedToken(XContentParser.Token.START_OBJECT, parser.currentToken(), parser);
            
            // parse current entity attribute pair
            while (parser.nextToken() != Token.END_OBJECT) {
                String fieldName = parser.currentName();
                // move to the field value
                parser.nextToken();
                switch (fieldName) {
                    case ATTRIBUTE_NAME_FIELD:
                        parsedName = parser.text();
                        break;
                    case ATTRIBUTE_VALUE_FIELD:
                        parsedValue = parser.text();
                    default:
                        break;
               }
               if (parsedName != null && parsedValue != null) {
                   entities.put(parsedName, parsedValue);
                   parsedName = null;
                   parsedValue = null;
               }
            }
        }
```